### PR TITLE
Dataset access control

### PR DESCRIPTION
**Pull request by ddjnw1yu**
- merge /program and /project api -> /access/scope
- create /access/token/<email> api
  - get policies from user.yaml file
- create and secure new /access/scope api
- create auth/jwt middleware
- update http status code
- add access control for graphql related api (query&pagination)
- update and add related testing
- update pagination structure
- update sgqlc schema
- optimize data fetching methods and logic
- split class
- optimize pagination payload

### CODE REVIEW

File: app/main.py [hunk 1]
```diff
@@ -142,55 +144,67 @@ async def root():
...
-async def get_gen3_program():
-    """
-    Return all programs information from the Gen3 Data Commons.
-    """
-    try:
-        program = SUBMISSION.get_programs()
-        return get_name_list(program, "program", "/v0/submission/")
-    except Exception as e:
-        raise HTTPException(status_code=NOT_FOUND, detail=str(e))
+@ app.get("/access/token/{email}", tags=["Gen3"], summary="Get gen3 access token for authorized user", responses=access_token_responses)
```
Comment (alan-wu) [inline app/main.py]: I think we should use a post request here instead.

File: middleware/auth.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+
+class Authenticator:
+    def __init__(self):
+        self.key = Fernet.generate_key()
+        self.fernet = Fernet(self.key)
+        self.access_token = None
+        self.authorized_email = None
+        self.public_access = False
+        self.current_user = None
```
Comment (alan-wu) [inline middleware/auth.py]: Can this class work with multiple users at the same time?

**Review by alan-wu — COMMENTED** [review]
A question about the authenticator class and probably a good idea to turn /access/token endpoint to a post request.

File: middleware/auth.py (hunk 2 above)
Comment (ddjnw1yu) [reply middleware/auth.py]: I have thought about this question. Haven't tested yet, probably not support. Will take a look at this.

**Review by alan-wu — APPROVED** [review]
Looks good to me.

**Review by alan-wu — COMMENTED** [review]
@ddjnw1yu A quick question - do you think there will be an issue with calling SESSION.cleanup while there is an active ongoing task  such as downloading?

**Review by alan-wu — APPROVED** [review]
This is working as expected.

**Pull request merged**